### PR TITLE
feat: GATK utils for common IO files

### PR DESCRIPTION
**Pull request by tdayris**
Utilities for bam/vcf indexes, for argument file(s) and MD5 files.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new function to construct GATK command-line options based on user-defined inputs and parameters in Snakemake workflows.
- **Bug Fixes**
	- Improved error handling to prevent conflicts in command-line options when multiple parameters are provided.

### CODE REVIEW

**Review by fgvieira — APPROVED** [review]

**Comment by fgvieira** [comment]
Thanks for the contribution!
I was just wondering if it would make more sense to have the defaults to `True` or `False`? I know the others (e.g. `samtools`) default to `True`, but it might make more sense to always default to `False` so that the parsing of the option is specifically activated. What do you think?

**Comment by tdayris** [comment]
I see what you mean, however, as these command line arguments are triggered by output files, I thought we would assume that user knows what they are doing and won't request files a GATK subcommand can't produce. Shall I turn these values to `False` ?

**Comment by fgvieira** [comment]
It just feels a bit safer to have everything disabled by default so that each wrapper has to specifically activate the options it needs.
But do not have a strong opinion, so up to you! :smile:

File: snakemake_wrapper_utils/gatk.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            gatk_opts += f" --reference {ref} "
+
+    if parse_ref_dict:
+        if is_arg("--sequence-dictionary", extra):
+            sys.exit(
+                "You have specified reference sequence dictionary (`--sequence-dictionary`) in `params.extra`; this is automatically inferred from `input.ref_dict`."
+            )
+        ref_dict = snakemake.input.get("ref_dict", "")
+        if ref_dict:
+            gatk_opts += f" --sequence-dictionary {ref_dict} "
```
Comment (fgvieira) [inline snakemake_wrapper_utils/gatk.py]: Just for consistency with current wrappers.
```suggestion
                "You have specified reference sequence dictionary (`--sequence-dictionary`) in `params.extra`; this is automatically inferred from `input.dict`."
            )
        dict = snakemake.input.get("dict", "")
        if dict:
            gatk_opts += f" --sequence-dictionary {dict}"
```

File: snakemake_wrapper_utils/gatk.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    ######################
+
+    if parse_ref:
+        if is_arg("-R", extra) or is_arg("--reference", extra):
+            sys.exit(
+                "You have specified reference file (`-R,--reference`) in `params.extra`; this is automatically inferred from `input.ref`."
+            )
+        ref = snakemake.input.get("ref")
+        if ref:
+            gatk_opts += f" --reference {ref} "
```
Comment (fgvieira) [inline snakemake_wrapper_utils/gatk.py]:
```suggestion
            gatk_opts += f" --reference {ref}"
```

File: snakemake_wrapper_utils/gatk.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    ###########################
+    ### Optional BAM output ###
+    ###########################
+    if parse_bam_index:
+        if is_arg("--create-output-bam-index", extra) or is_arg("-OBI", extra):
+            sys.exit(
+                "You have specified bam index creation (`-OBI,--create-output-bam-index`) in `params.extra`; this is automatically inferred from `output.bam_bai`."
+            )
+        if snakemake.output.get("bam_bai"):
+            gatk_opts += " --create-output-bam-index "
```
Comment (fgvieira) [inline snakemake_wrapper_utils/gatk.py]:
```suggestion
            gatk_opts += " --create-output-bam-index"
```

File: snakemake_wrapper_utils/gatk.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        if snakemake.output.get("bam_bai"):
+            gatk_opts += " --create-output-bam-index"
+
+    if parse_bam_md5:
+        if is_arg("--create-output-bam-md5", extra) or is_arg("-OBM", extra):
+            sys.exit(
+                "You have specified bam MD5-sum creation (`-OBM,--create-output-bam-md5`) in `params.extra`; this is automatically inferred from `output.bam_md5`."
+            )
+        if snakemake.output.get("bam_md5"):
+            gatk_opts += " --create-output-bam-md5 "
```
Comment (fgvieira) [inline snakemake_wrapper_utils/gatk.py]:
```suggestion
            gatk_opts += " --create-output-bam-md5"
```

**Pull request merged**